testgrid/cmd/updater/main.go

```diff
@@ -631,82 +1059,133 @@ func Group(cfg config.Configuration, name string) (*config.TestGroup, bool) {
+	var wg sync.WaitGroup
+
+	for i := uint(0); i < opt.groupConcurrency; i++ {
+		wg.Add(1)
+		go func() {
+			for tg := range groups {
+				if err := updateGroup(client, ctx, tg, opt.config.testGroup(tg.Name), opt.buildConcurrency, opt.confirm); err != nil {
+					log.Printf("FAIL: %v", err)
+				}
+			}
+			wg.Done()
+		}()
 	}
-	log.Println(tg)
 
-	bkt := client.Bucket("kubernetes-jenkins")
-	attrs, err := bkt.Attrs(ctx)
-	if err != nil {
-		log.Fatalf("Failed to access bucket: %v", err)
+	if opt.group != "" { // Just a specific group
+		// o := "ci-kubernetes-test-go"
+		// o = "ci-kubernetes-node-kubelet-stable3"
+		// gs://kubernetes-jenkins/logs/ci-kubernetes-test-go
+		// gs://kubernetes-jenkins/pr-logs/pull-ingress-gce-e2e
+		o := opt.group
+		if tg, ok := Group(*cfg, o); !ok {
+			log.Fatalf("Failed to find %s in %s", o, opt.config)
+		} else {
+			groups <- *tg
+		}
+	} else { // All groups
+		for _, tg := range cfg.TestGroups {
+			groups <- *tg
+		}
 	}
-	fmt.Printf("bucket %s, attrs %v", bkt, attrs)
+	close(groups)
+	wg.Wait()
+}
+
+func updateGroup(client *storage.Client, ctx context.Context, tg config.TestGroup, gridPath gcsPath, concurrency uint, write bool) error {
+	o := tg.Name
+
+	var tgPath gcsPath
+	if err := tgPath.Set("gs://" + tg.GcsPrefix); err != nil {
+		return fmt.Errorf("group %s has an invalid gcs_prefix %s: %v", o, tg.GcsPrefix, err)
+	}
+
 	g := state.Grid{}
 	g.Columns = append(g.Columns, &state.Column{Build: "first", Started: 1})
-	fmt.Println(g)
-	builds := make(chan Build)
-	uPR := "gs://kubernetes-jenkins/pr-logs/pull-ingress-gce-e2e"
-	uCI := "gs://kubernetes-jenkins/logs/ci-kubernetes-test-go"
-	uCIN := "gs://kubernetes-jenkins/logs/ci-kubernetes-node-kubelet-stable3"
-	_ = uCI
-	_ = uPR
-	u := uCIN
-	go func() {
-		if err = ListBuilds(client, ctx, u, builds); err != nil {
-			log.Fatalf("Failed to list builds: %v", err)
+	builds, err := listBuilds(client, ctx, tgPath)
+	if err != nil {
+		return fmt.Errorf("failed to list %s builds: %v", o, err)
+	}
+	grid, err := ReadBuilds(ctx, tg, builds, 50, Days(7), concurrency)
+	if err != nil {
+		return err
+	}
+	buf, crc, err := marshalGrid(*grid)
+	if err != nil {
+		return fmt.Errorf("failed to marshal %s grid: %v", o, err)
+	}
+	tgp := gridPath
+	if !write {
+		log.Printf("  Not writing %s (%d bytes) to %s", o, len(buf), tgp)
+	} else {
+		log.Printf("  Writing %s (%d bytes) to %s", o, len(buf), tgp)
+		if err := uploadBytes(client, ctx, tgp, buf, crc); err != nil {
+			return fmt.Errorf("upload %s to %s failed: %v", o, tgp, err)
 		}
-		close(builds)
-	}()
-	grid := ReadBuilds(*tg, builds, 1000, Days(5))
-	log.Printf("Grid: %d %s", len(grid.Columns), grid.String())
+	}
+	log.Printf("WROTE: %s, %dx%d grid (%s, %d bytes)", tg.Name, len(grid.Columns), len(grid.Rows), tgp, len(buf))
+	return nil
+}
+
+// marhshalGrid serializes a state proto into zlib-compressed bytes and its crc32 checksum.
+func marshalGrid(grid state.Grid) ([]byte, uint32, error) {
```

Comment:
Why does this return a CRC? The only thing that cares about the CRC is uploadBytes, so it should continue to calculate it itself.

Reply:
Done